internal/routes/all_routes.go

```diff
@@ -23,15 +24,21 @@ func SetupRoutes(r *chi.Mux) {
 	r.Get("/ping", s.StatusService)
 	r.Route("/docs", func(r chi.Router) {
 		r.Use(redocMiddleware)
-		r.Get("/openapi.json", api.ServeOpenAPISpec)
+		r.Route("/openapi.json", func(r chi.Router) {
+			r.Use(middleware.ETagMiddleware(15*time.Minute, "json-spec", api.ETagJSONSpec))
+			r.Get("/", api.ServeOpenAPISpec)
+		})
 	})
 	r.Mount(PathPrefix(), apiRouter())
 }
 
 func apiRouter() http.Handler {
 	r := chi.NewRouter()
 
-	r.Get("/openapi.json", api.ServeOpenAPISpec)
+	r.Route("/openapi.json", func(r chi.Router) {
+		r.Use(middleware.ETagMiddleware(15*time.Minute, "json-spec", api.ETagJSONSpec))
+		r.Get("/", api.ServeOpenAPISpec)
+	})
```

Comment:
And this is how to use the middleware, I set the timeout to 15 minutes. I use it both for JSON spec and documentation.

It's not super useful for OpenAPI, granted, but it's a good example. A good candidate will definitely be instance-types endpoint.